### PR TITLE
Adding Adaptive SpecAugment

### PR DESCRIPTION
**Pull request by ShampooWang**
#236 Hi! Me and @cornliu have implemented Adaptive SpecAugment for downstream tasks (ASR & CTC) and pretrain tasks by modifying  several files.

### CODE REVIEW

**Comment by andi611** [comment]
@ShampooWang @cornliu Thanks! This is sufficient for tomorrow's project deadline (we do not have to merge this by tomorrow).
It will be nice to have some comparison results on using "spec aug" v.s. "adaptive spec aug" for tomorrow's presentation. Great work!

**Comment by andi611** [comment]
@leo19941227 Can you check if modifying these config files are appropriate for the **launched SUPERB v1** challenge? Or we can do a separate config file, without modifying the original configs.

**Comment by leo19941227** [comment]
Thanks @ShampooWang @cornliu and @andi611!

I find that this PR will change the default behavior of ASR and SNIPS to use adaptive specaug. I think it will be nice we make sure the new configs have the exactly same behavior compared to the original config.

**Comment by cornliu** [comment]
We change the adaptive default value from true to false in the yaml file. It may let the new configs have the same behavior with the old one.

**Review by leo19941227 — COMMENTED** [review]
Thanks for the contribution! I have a few minor comments.

File: s3prl/downstream/specaug.py [hunk 1]
```diff
@@ -208,10 +227,23 @@ def mask_along_axis(self, spec, spec_lengths):
         B = spec.shape[0]
         # D = Length or Freq
         D = spec.shape[self.dim]
+        T = self.mask_width_range[1]
+        # Adaptive_SpecAugment
+        if self.dim == 1 & self.adaptive :
+          if self.adaptive_number_ratio > 0:
+            self.num_mask = min(int(self.adaptive_number_ratio * D), self.max_n_time_masks)
+          else:
+            self.num_mask = num_mask
```
[the remark this answers is not in the document]
Comment (ShampooWang) [reply s3prl/downstream/specaug.py:236]: Do you mean line 235-236? It seems like line 235-236 is redundant, you can just simply remove it. I just want to make sure that adaptive number ratio is greater than zero, otherwise we do not change the number of masks. Also, I am a little confused about the difference between `self.num_mask `and `num_mask` as you mentioned below.

File: s3prl/downstream/specaug.py [hunk 2]
```diff
@@ -208,10 +227,23 @@ def mask_along_axis(self, spec, spec_lengths):
         B = spec.shape[0]
         # D = Length or Freq
         D = spec.shape[self.dim]
+        T = self.mask_width_range[1]
+        # Adaptive_SpecAugment
+        if self.dim == 1 & self.adaptive :
+          if self.adaptive_number_ratio > 0:
+            self.num_mask = min(int(self.adaptive_number_ratio * D), self.max_n_time_masks)
```
[the remark this answers is not in the document]
Comment (ShampooWang) [reply s3prl/downstream/specaug.py:234]: Yes, I think we should change the code as your suggestion, since I may misconstrue the difference between the `self.num_mask` and `num_mask`. Thanks for your correction!

**Comment by ShampooWang** [comment]
@leo19941227 and @andi611 May I ask which version of time wrapping did `class TimeWarp` follow from? I assumed that your implementation followed from the [original version](https://arxiv.org/pdf/1904.08779.pdf) or the [second version](https://arxiv.org/pdf/1912.05533.pdf), whose author provided a specific formula. Although two versions are equivalent, I'm just afraid that if your `class TimeWarp` followed from other versions of time wrapping, and unfortunately it is not equivalent to the versions as I mentioned above, then maybe we should modify `class TimeWarp`.

**Comment by leo19941227** [comment]
We follow the ESPnet experience where they find a few large masks are better than lots of small masks.

https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/conf/tuning/train_asr_conformer6_n_fft512_hop_length256.yaml

**Comment by leo19941227** [comment]
We follow the ESPnet implementation here: https://espnet.github.io/espnet/_modules/espnet2/layers/time_warp.html
I think we can always provide new classes to support different implementations instead of modifying the existing one.
Also, I am curious about how much this difference affect the final performance. It will be nice to run some experiments.

**Comment by ShampooWang** [comment]
I found that [ESPnet's](https://aclanthology.org/2020.acl-demos.34.pdf) implementation of SpecAugment followed from this [paper](https://arxiv.org/pdf/1911.08876.pdf), which followed from the [original version](https://arxiv.org/pdf/1904.08779.pdf) as I mentioned above, so it is equivalent to the second version, my main reference. Also, I also found that the author of SpecAugment have mentioned that "**Time warping contributes, but is not a major factor in improving performance**", "**Time warping,
being the most expensive as well as the least influential of the
augmentations discussed in this work."**. In conclusion, I think we don't have to modify `class TimeWarp`.

**Comment by ShampooWang** [comment]
However, I am also curious about the possibility of some differences, so I am willing to provide a new class of `TimeWarp`, an implementation of my [main reference](https://arxiv.org/pdf/1912.05533.pdf), but I think this implementation is more appropriate to be requested in a new PR

**Comment by leo19941227** [comment]
Sure! Please go ahead. I look forward to the new PR.
Please also consider to add some results (ablation tables) for the newly added functions in the PR page. Thanks!

**Comment by ShampooWang** [comment]
Sorry for my delay, it took me 5-6 days to run over 200k stpes, and the learning curves are presented in this picture. 
s3prl_asr_experiment
We found some interest results. First, the training loss of Adaptive SpecAugment was way more higher than the default. Second, Adaptive SpecAugment needs more steps to converge. Last but not least, the **wer** of the **default** on `dev-clean` is **23.28**(which is differ from **23.18** mentioned in your paper. Do I have to run more steps or adjust some hyper-parameters to achieve?), and  the **wer** of the **Adaptive SpecAugment** on `dev-clean` is **23.3**. I think the performance was effected by the small size of mask windows, and I infer that if we take more steps in training or adjust some hyper parameters, Adaptive SpecAugment will outperform the SpecAugment.

File: s3prl/downstream/specaug.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ShampooWang) [reply s3prl/downstream/specaug.py]: > It is a little bit weird to me that an attribute (`self.num_mask`) is dynamically changed along with the most recent input. Could we change to `num_mask = min(int(self.adaptive_number_ratio * D), self.max_n_time_masks)` and use `num_mask` (instead of `self.num_mask`) in the code below?

I have followed you recommendation, using variables instead of directly changing `self.num_mask`.

**Comment by ShampooWang** [comment]
After my evaluation, I decided to not change pretrain's SpecAugment to adpative, so I change the pretrain part to your original version, or you can just block the merging of any file changes under s3prl/s3prl/pretrain/spec_augment/ For downstream parts, I think it is ready to be merged, if everything is ok, I will close this PR.

**Comment by leo19941227** [comment]
Hi @ShampooWang , 

For the 0.1 WER difference, I think it is not significant, especially when CTC loss implemented in PyTorch is essentially non-deterministic by default. (You can make it deterministic by sacrificing the training speed)
Hence, even with the exactly same seed, you can't get the exactly same result.
According to my experience, this difference is acceptable.

If you are interested in what is the outcome of more training steps, you don't need to re-train it.
You can simply load the checkpoint back and continuing the training by overriding the `total_steps`.
Specifically:

```
python3 run_downstream.py -e `your_last_checkpoint.ckpt` -o config.runner.total_steps=300000
```

Then, the training will continue for the extra 100k steps, and the all the artifacts will appear in the same `expdir` you previously assign (you can also override this by `-o args.expdir=newpath`), including the new Tensorboard event file. Then, you can see the training curves grow from where it stopped, which is 200k steps. Just in case you are interested.

May I ask about you plan on closing the PR? Do you wish to merge this PR or open another PR for the adaptive-specaugment?

**Comment by ShampooWang** [comment]
Very appreciate to your recommendation! I was wondering how to run more steps but not rerun the whole experiment. Since I am a novice, I misconstrue the function of "closing PR", what I really wish is to merge this PR and end this topic.

**Comment by leo19941227** [comment]
Hi @ShampooWang and @cornliu ,

Thanks for the contribution. LGTM!
I am merging this.

**Review by leo19941227 — APPROVED** [review]
